### ISSUE TITLE
Value never used warning is wrongly thrown for named return variables

### ISSUE THREAD

**Issue by shahrk** [golangci/golangci-lint#1294]
Thank you for creating the issue!

- [x] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [x] Yes, I've searched similar issues on GitHub and didn't find any.
- [x] Yes, I've included all information below (version, config, etc).

As the named return variable is actually being returned & thus used, the warning being thrown is wrong.
```
func Hello() (err error) {
    err = errorThrowingFunc()
    return
}
```

Please include the following information:

<details><summary>Version of golangci-lint</summary>

```console
$ golangci-lint --version
golangci-lint has version 1.30.0 built from 45b90f6 on 2020-08-03T03:10:34Z
```

</details>

<details><summary>Config file</summary>

```console
$ cat .golangci.yml
cat: .golangci.yml: No such file or directory
```

</details>

<details><summary>Go environment</summary>

```console
$ go version && go env
go version go1.13.8 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/root/.cache/go-build"
GOENV="/root/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/src/go-workspace"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/lib/go-1.13"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/go-1.13/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/src/go-workspace/src/shahrk/rpc/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build623854947=/tmp/go-build -gno-record-gcc-switches"
```

</details>

<details><summary>Verbose output of running</summary>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /src/go-workspace/src/shahrk/rpc /src/go-workspace/src/shahrk /src/go-workspace/src /src/go-workspace /src /]
INFO [lintersdb] Active 10 linters: [deadcode errcheck gosimple govet ineffassign staticcheck structcheck typecheck unused varcheck]
INFO [loader] Go packages loading at mode 575 (types_sizes|compiled_files|exports_file|files|name|deps|imports) took 224.4965ms
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 5.6662ms
INFO [linters context/goanalysis] analyzers took 20.6068ms with top 10 stages: buildir: 7.1212ms, unreachable: 3.2549ms, isgenerated: 2.4146ms, ineffassign: 2.2036ms, fact_deprecated: 324µs, SA3001: 261.3µs, copylocks: 249µs, atomic: 204.2µs, ctrlflow: 200.2µs, SA1006: 198.4µs
INFO [linters context/goanalysis] analyzers took 8.4021ms with top 10 stages: buildir: 7.8509ms, U1000: 551.2µs
INFO [runner] Issues before processing: 2, after processing: 1
INFO [runner] Processors filtering stat (out/in): max_same_issues: 1/1, autogenerated_exclude: 2/2, max_per_file_from_linter: 1/1, max_from_linter: 1/1, severity-rules: 1/1, filename_unadjuster: 2/2, skip_files: 2/2, identifier_marker: 2/2, exclude-rules: 1/1, nolint: 1/1, source_code: 1/1, path_shortener: 1/1, path_prefixer: 1/1, cgo: 2/2, skip_dirs: 2/2, exclude: 1/2, uniq_by_line: 1/1, diff: 1/1, sort_results: 1/1, path_prettifier: 2/2
INFO [runner] processing took 6.5233ms with stages: autogenerated_exclude: 2.4211ms, source_code: 2.192ms, nolint: 1.6989ms, exclude: 67.3µs, path_prettifier: 61.1µs, identifier_marker: 34µs, skip_dirs: 30.7µs, cgo: 2.7µs, uniq_by_line: 2.7µs, max_same_issues: 2.5µs, path_shortener: 1.9µs, filename_unadjuster: 1.6µs, max_from_linter: 1.4µs, max_per_file_from_linter: 900ns, diff: 900ns, skip_files: 800ns, exclude-rules: 800ns, sort_results: 800ns, severity-rules: 700ns, path_prefixer: 500ns
INFO [runner] linters took 243.1765ms with stages: goanalysis_metalinter: 151.9692ms, unused: 84.463ms
server/server.go:72:2: SA4006: this value of `err` is never used (staticcheck)
        err = rpc.Register(vendy)
        ^
INFO File cache stats: 1 entries of total size 1.5KiB
INFO Memory: 6 samples, avg is 69.5MB, max is 69.6MB
INFO Execution took 486.7555ms
```

</details>

**Comment by denis-tingajkin** [comment]
Interesting... Can you check is this issue reproducible with pure staticcheck to understand that the problem related to golangci-lint?

**Comment by bombsimon** [comment]
You verbose output indicate that it's ran on different code than your example code. I cannot reproduce with the sample you provided.

```sh
$ golangci-lint version
golangci-lint has version v1.30.0 built from (unknown, mod sum: "h1:UhdK5WbO0GBd7W+k2lOD7BEJH4Wsa7zKfw8m3/aEJGQ=") on (unknown)

$ cat *.go
package x

import "errors"

func Hello() (err error) {
        err = errors.New("X")
        return
}

$ golangci-lint run -v --no-config --disable-all --enable=staticcheck ./.
INFO [lintersdb] Active 1 linters: [staticcheck]
INFO [loader] Go packages loading at mode 575 (types_sizes|deps|exports_file|imports|compiled_files|files|name) took 132.575344ms
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 187.857µs
INFO [linters context/goanalysis] analyzers took 0s with no stages
INFO [runner] processing took 5.738µs with stages: nolint: 2.163µs, max_same_issues: 762ns, skip_dirs: 355ns, max_from_linter: 252ns, path_prettifier: 197ns, cgo: 196ns, diff: 183ns, path_shortener: 179ns, uniq_by_line: 170ns, source_code: 169ns, filename_unadjuster: 169ns, autogenerated_exclude: 164ns, skip_files: 163ns, identifier_marker: 161ns, severity-rules: 89ns, path_prefixer: 79ns, max_per_file_from_linter: 75ns, exclude-rules: 72ns, sort_results: 72ns, exclude: 68ns
INFO [runner] linters took 15.907206ms with stages: staticcheck: 15.842265ms
INFO File cache stats: 0 entries of total size 0B
INFO Memory: 3 samples, avg is 70.7MB, max is 70.8MB
INFO Execution took 162.167828ms

$ echo $?
0
```

Could you provide the code from `server/server.go:72` that actually generates this message?

**Comment by shahrk** [comment]
Yeah I just provided a sample, following is the actual code (the warning is thrown for the _3rd line of the Start() method_):
```
package server

import (
	"fmt"
	"log"
	"net"
	"net/http"
	"net/rpc"
	"sync"
)

type GetArgs struct {
	Key string
}

type GetReply struct {
	Err   string
	Value string
}

type PutArgs struct {
	Key, Value string
}

type PutReply struct {
	Err string
}

// Vendy is a thread-safe key value store
type Vendy struct {
	store map[string]string
	mux   sync.Mutex
}

func (v *Vendy) Get(args *GetArgs, reply *GetReply) (err error) {
	v.mux.Lock()
	defer v.mux.Unlock()
	if val, exists := v.store[args.Key]; !exists {
		log.Println("Unfulfilled GET Request")
		reply.Err = "Requested key does not exist"
	} else {
		log.Println("Fulfilled GET Request on Vendy")
		reply.Value = val
	}
	return
}

func (v *Vendy) Put(args *PutArgs, reply *PutReply) (err error) {
	v.mux.Lock()
	defer v.mux.Unlock()
	v.store[args.Key] = args.Value
	log.Println("Fulfilled PUT Request on Vendy")
	return
}

type Server struct {
	Port     uint
	listener net.Listener
}

func (server *Server) Stop() (err error) {
	if server.listener != nil {
		log.Println("GOODBYE")
		server.listener.Close()
	}
	return
}

func (server *Server) Start() (err error) {
	vendy := new(Vendy)
	vendy.store = make(map[string]string)
	err = rpc.Register(vendy)
	rpc.HandleHTTP()
	server.listener, err = net.Listen("tcp", fmt.Sprint(":", server.Port))

	if err != nil {
		log.Fatal("Listen error:", err)
		return
	} else {
		log.Println("Server listening on port ", server.Port)
	}
	go func() {
		if err := http.Serve(server.listener, nil); err != nil {
			log.Fatal("Serve error:", err)
		}
	}()
	return
}
```

**Comment by bombsimon** [comment]
It's because you assign err on line 3, then you never check for it and overwrite it at line 5 again. This will reproduce the same issue:

```go
package x

import (
	"errors"
	"fmt"
)

func Hello() (err error) {
	err = errors.New("X")

	fmt.Println("Ignoring err!")

	err = errors.New("another error")

	return
}
```

```sh
golangci-lint run -v --no-config --disable-all --enable=staticcheck
INFO [lintersdb] Active 1 linters: [staticcheck]
INFO [loader] Go packages loading at mode 575 (files|name|exports_file|deps|imports|types_sizes|compiled_files) took 158.038184ms
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 203.001µs
INFO [linters context/goanalysis] analyzers took 381.388372ms with top 10 stages: buildir: 244.579322ms, fact_purity: 106.037728ms, fact_deprecated: 28.886902ms, SA4015: 194.855µs, SA1026: 190.706µs, SA4009: 190.549µs, SA1021: 184.846µs, SA1003: 169.924µs, SA1013: 75.894µs, isgenerated: 62.533µs
INFO [runner] Issues before processing: 2, after processing: 1
INFO [runner] Processors filtering stat (out/in): severity-rules: 1/1, cgo: 2/2, filename_unadjuster: 2/2, uniq_by_line: 1/2, source_code: 1/1, path_shortener: 1/1, path_prettifier: 2/2, autogenerated_exclude: 2/2, exclude-rules: 2/2, sort_results: 1/1, max_from_linter: 1/1, path_prefixer: 1/1, skip_files: 2/2, identifier_marker: 2/2, exclude: 2/2, nolint: 2/2, diff: 1/1, skip_dirs: 2/2, max_per_file_from_linter: 1/1, max_same_issues: 1/1
INFO [runner] processing took 280.481µs with stages: exclude: 72.526µs, nolint: 56.37µs, identifier_marker: 39.962µs, path_prettifier: 35.358µs, autogenerated_exclude: 30.049µs, source_code: 24.411µs, skip_dirs: 10.253µs, uniq_by_line: 2.631µs, cgo: 2.598µs, max_same_issues: 1.757µs, path_shortener: 1.113µs, filename_unadjuster: 825ns, max_from_linter: 721ns, max_per_file_from_linter: 547ns, exclude-rules: 290ns, diff: 285ns, skip_files: 252ns, severity-rules: 242ns, sort_results: 161ns, path_prefixer: 130ns
INFO [runner] linters took 330.327308ms with stages: staticcheck: 322.006827ms
main.go:9:2: SA4006: this value of `err` is never used (staticcheck)
        err = errors.New("X")
        ^
INFO File cache stats: 1 entries of total size 168B
INFO Memory: 7 samples, avg is 71.1MB, max is 71.3MB
INFO Execution took 507.069812ms
```

You should probably add this to line 4:

```go
if err != nul {
    return
}
```

**Comment by shahrk** [comment]
My bad 🤦‍♂️
